dvc/rwlock.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import unicode_literals
+
+import os
+import json
+
+from collections import defaultdict
+from contextlib import contextmanager
+
+from voluptuous import Schema, Invalid
+
+from .exceptions import DvcException
+from .lock import LockError
+from .utils.compat import (
+    convert_to_unicode,
+    FileNotFoundError,
+    JSONDecodeError,
+    str,
+)
+from .utils.fs import relpath
+
+INFO_SCHEMA = {"pid": int, "cmd": str}
+
+SCHEMA = Schema({"write": {str: INFO_SCHEMA}, "read": {str: [INFO_SCHEMA]}})
+
+
+class RWLockFileCorruptedError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileCorruptedError, self).__init__(
+            "Unable to read RWLock-file '{}'. JSON structure is "
+            "corrupted".format(relpath(path)),
+            cause=cause,
+        )
+
+
+class RWLockFileFormatError(DvcException):
+    def __init__(self, path, cause):
+        super(RWLockFileFormatError, self).__init__(
+            "RWLock-file '{}' format error.".format(relpath(path)), cause=cause
+        )
+
+
+@contextmanager
+def _edit_rwlock(lock_dir):
+    path = os.path.join(lock_dir, "rwlock")
+    try:
+        with open(path, "r") as fobj:
+            lock = json.load(fobj)
+        lock = SCHEMA(convert_to_unicode(lock))
+    except FileNotFoundError:
+        lock = SCHEMA({})
+    except JSONDecodeError as exc:
+        raise RWLockFileCorruptedError(path, cause=exc)
+    except Invalid as exc:
+        raise RWLockFileFormatError(path, cause=exc)
+    lock = defaultdict(dict, lock)
+    lock["read"] = defaultdict(list, lock["read"])
+    lock["write"] = defaultdict(dict, lock["write"])
+    yield lock
+    with open(path, "w+") as fobj:
+        json.dump(lock, fobj)
+
+
+def _info_to_str(info):
+    return "  (PID {}): {}".format(info["pid"], info["cmd"])
+
+
+def _get_urls(lock, path_info):
+    ret = set(filter(path_info.overlaps, lock.keys()))
+    ret.add(path_info.url)
+    return ret
+
+
+def _check_no_writers(lock, info, path_infos):
+    for path_info in path_infos:
+        urls = _get_urls(lock["write"], path_info)
+        writers = [lock["write"].get(url) for url in urls]
+        writers = list(filter(bool, writers))
+        if not writers:
+            continue
+
+        assert len(writers) == 1
+        writer = writers[0]
+        if writer == info:
+            continue
+
+        raise LockError(
+            "'{}' is busy, it is being written to by:\n{}".format(
+                str(path_info), _info_to_str(writer)
+            )
+        )
+
+
+def _check_no_readers(lock, path_infos):
+    from itertools import chain
+
+    for path_info in path_infos:
+        urls = _get_urls(lock["read"], path_info)
+
+        readers = list(
+            chain.from_iterable([lock["read"].get(url, []) for url in urls])
+        )
```

Comment:
Same as above, we don't need to multiply `path_info` with `_get_urls()`. We should just:

```python
blocking_urls = lfilter(path_info.overlaps, lock['read'])

# then if we allow empty lists in lock['read']
blocking_procs = lmapcat(lock['read'].get, blocking_urls)
if blocking_procs:
    # ...

# if we don't allow empty lists in lock['read']
if blocking_urls:
    blocking_procs = lmapcat(lock['read'].get, blocking_urls)
    # ...
```

I am not pushing for `cat/lcat/lmapcat` I simply write my suggestions, these two lines might be written as:

```python
blocking_urls = list(filter(path_info.overlaps, lock['read']))
blocking_procs = list(chain.from_iterable(map(lock['read'].get, blocking_urls)))
```
Or mixed with some generator expressions.

Reply:
@Suor Sorry, my previous message might've sounded too harsh. These tricks are neat, just takes time to think in terms of funcy helpers 🙂